pkg/challenge/response.go

```diff
@@ -29,23 +30,47 @@ func (o *OuterResponse) Open(privateEncryptionKey [32]byte) (*InnerResponse, err
 		return nil, fmt.Errorf("unmarshaling inner box: %w", err)
 	}
 
-	counterPartyPubKey, err := echelper.PublicPemToEcdsaKey(innerResponse.PublicSigningKey)
-	if err != nil {
-		return nil, fmt.Errorf("unmarshalling public ecdsa signing key from pem: %w", err)
+	if err := verifyWithKeyBytes(innerResponse.PublicSigningKey, innerResponseBytes, o.Sig); err != nil {
+		return nil, fmt.Errorf("verifying challenge signature: %w", err)
+	}
+
+	// neither signature 2 nor public signing key 2 was provided, return what we have
+	if (o.Sig2 == nil || len(o.Sig2) <= 0) && (innerResponse.PublicSigningKey2 == nil || len(innerResponse.PublicSigningKey2) <= 0) {
+		return &innerResponse, nil
+	}
```

Comment:
I'm not sure if this should require both to be nil, or allow the sig to be nil, even if the inner is present.

I'm thinking about the timeout case.
1. The inner is constructed with both public keys, then it's handed to both signers. 
2. The TPM signer times out. 
3. Now it's all invalid

Reply:
yea, I think your correct, just check for nil sig

Reply:
updated to just check sig2

Reply:
Add tests :P

Reply:
Hrm. We could, maybe, strip key2 out of the inner, after verifying it? That would ensure the challenger doesn't mis use it

Reply:
> Add tests :P

There are tests here that use nil signer 2

https://github.com/James-Pickett/krypto/blob/james/sig2/pkg/challenge/challenge_test.go#L81
https://github.com/James-Pickett/krypto/blob/james/sig2/test/krypto/challenge_test.rb#L31


> Hrm. We could, maybe, strip key2 out of the inner, after verifying it? That would ensure the challenger doesn't mis use it

Doesn't the challenger need the key to look up the device?

Reply:
I probably didn't use enough words. I think we probably need some convoluted logic.

The responder has 2 keys. So it needs to add them both to inner. Then it signs inner. But some of those signatures might timeout. The net effect here is a single signature. (But we don't want to regenerate inner here, since it would invalidate the signature)

Then, the challengers receives ^^^. If it's got two sigs, it's easy. They both need to pass. But, if there's only 1 signature... That signature needs to pass, easy. But the second key inside inner cannot be verified. So something needs to know not to use it. My inclination is to stash that logic in `response.Open`. It verifies the signatures, and if there's no sig2, then it removes key2 lest the caller try to use it

Reply:
> I probably didn't use enough words. I think we probably need some convoluted logic.
> 
> The responder has 2 keys. So it needs to add them both to inner. Then it signs inner. But some of those signatures might timeout. The net effect here is a single signature. (But we don't want to regenerate inner here, since it would invalidate the signature)
> 
> Then, the challengers receives ^^^. If it's got two sigs, it's easy. They both need to pass. But, if there's only 1 signature... That signature needs to pass, easy. But the second key inside inner cannot be verified. So something needs to know not to use it. My inclination is to stash that logic in `response.Open`. It verifies the signatures, and if there's no sig2, then it removes key2 lest the caller try to use it

Cool, this makes sense, I will make it so!